src/main/java/frc/robot/subsystems/SwerveSubsystem.java

```diff
@@ -409,4 +429,22 @@ public void addFakeVisionReading() {
     swerveDrive.addVisionMeasurement(
         new Pose2d(3, 3, Rotation2d.fromDegrees(65)), Timer.getFPGATimestamp());
   }
+
+  public void updateFromLimelight() {
+    Optional<Alliance> team = DriverStation.getAlliance();
+    boolean hasTarget = LimelightHelpers.getTV(Constants.limelightName);
+    SmartDashboard.putBoolean("TargetResults Valid", hasTarget);
+    if (team.get() == Alliance.Red && hasTarget) {
+      swerveDrive.addVisionMeasurement(
```

Comment:
this needs fixing. the .times(-1) is done in different places for red or blue (setgyro vs getBotPose2d).

and while you're fixing, if you could  re-work it so